cpp_utils/include/cpp_utils/collection/database/SafeDatabase.hpp

```diff
@@ -0,0 +1,175 @@
+};
+
+/**
+ * This class implements the Interface \c IModificableDatabase in a thread safe way.
+ *
+ * This represents a map of keys and values giving the methods require by IDatabase including modify and erase.
+ * It uses an internal std::map to store the data.
+ *
+ * The iteration over the internal values is thread safe.
+ * This means that while there is an alive iterator, the database could not change it state (add, modify, erase).
+ * However, it could still be asked for values or iterate from somewhere else.
+ *
+ * @motivation this class is required to instantiate the interface \c IModificableDatabase and it also represents
+ * a thread safe std::map.
+ *
+ * It adds the method \c at , \c size and some other utilities along with those of the interface.
+ *
+ * @tparam \c Key type to use as key/index of map.
+ * @tparam \c Value type to use as internal value stored on map.
+ */
+template <typename Key, typename Value>
+class SafeDatabase : public IModificableDatabase<Key, Value, SafeDatabaseIterator<Key, Value>>
+{
+public:
+
+    bool add(
+            Key&& key,
+            Value&& value) override;
+
+    bool modify(
+            const Key& key,
+            Value&& value) override;
+
+    bool erase(
+            const Key& key) override;
+
+    bool is(
+            const Key& key) const override;
+
+    SafeDatabaseIterator<Key, Value> find(
+            const Key& key) const override;
+
+    SafeDatabaseIterator<Key, Value> begin() const override;
+    SafeDatabaseIterator<Key, Value> end() const override;
+
+
+    /**
+     * @brief Add using copy semantics instead of movement.
+     *
+     * This is useful for those types that do not safe time moving (as native types) or when the scope of the variables
+     * to add to the database do not allow movement (const values).
+     */
+    bool add(
+            const Key& key,
+            const Value& value);
+
+    /**
+     * @brief Return a copy of the value indexed by \c key .
+     *
+     * Efficient call to return a copy of the internal value stored under \c key index.
+     * It throws in case the key is not present.
+     *
+     * @attention This method is only instantiated for those types that could be copied.
+     *
+     * @tparam V This must be equal to Database Value type.
+     * @warning V must be equal to Database Value type.
+     *
+     * @param key index of the value to look for.
+     *
+     * @return copy of the internal value if exist.
+     * @throw \c std::out_of_range if key not in database.
+     */
+    Value at(
+            const Key& key) const;
+
+    //! Number of keys stored.
+    unsigned int size() const noexcept;
+
+    /**
+     * @brief Add a new value or modify an existent one with the new value.
+     *
+     * @param key key indexing the new or existent value.
+     * @param value new value to set.
+     *
+     * @return true if value has been added.
+     * @return false if value already exists, in this case the value is modified.
```

Comment:
It could be confusing return false

Reply:
I know, but I do not know what else to do but add an enumeration with the possible return :thinking: . Not sure that is a good option either.